lib/pbench/server/api/resources/__init__.py

```diff
@@ -1567,27 +1597,88 @@ def _dispatch(
                 self._check_authorization(auth_params)
             except UnauthorizedAccess as e:
                 self.logger.warning("{}: {}", api_name, e)
+                if audit:
+                    Audit.create(
+                        root=audit,
+                        status=AuditStatus.FAILURE,
+                        reason=AuditReason.PERMISSION,
+                        attributes={"message": str(e)},
+                    )
                 abort(e.http_status, message=str(e))
             except Exception as e:
                 self.logger.exception("{}: {}", api_name, e)
+                if audit:
+                    Audit.create(
+                        root=audit,
+                        status=AuditStatus.FAILURE,
+                        reason=AuditReason.INTERNAL,
+                        attributes={"message": str(e)},
+                    )
                 abort(
                     HTTPStatus.INTERNAL_SERVER_ERROR,
                     message=HTTPStatus.INTERNAL_SERVER_ERROR.phrase,
                 )
 
+        # Pass the root audit object to the API implementation. Normally, we'll
+        # automatically generate an audit finalization; however if the API
+        # wants to emit a special audit sequence it can disable "finalize"
+        # in the context. It can also pass "attributes" by setting that
+        # field.
+        context = {
+            "auditing": {
+                "audit": audit,
+                "finalize": bool(audit),
+                "status": AuditStatus.SUCCESS,
+                "reason": None,
+                "attributes": None,
+            }
+        }
+
         try:
-            return execute(params, request)
+            response = execute(params, request, context)
+            auditing = context["auditing"]
+            if auditing["finalize"]:
+                Audit.create(
+                    root=auditing["audit"],
+                    status=auditing["status"],
+                    reason=auditing["reason"],
+                    attributes=auditing["attributes"],
+                )
```

Comment:
I'd have to `pop` `"finalize"` to do that, and I'm not sure it's worthwhile (since the actual payload becomes somewhat obfuscated in the process), but it's an interesting idea.